File: mobile/android/OpenU60/app/src/main/java/com/openu60/feature/router/lan/LANSettingsViewModel.kt
package com.openu60.feature.router.lan

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openu60.core.model.LANConfig
import com.openu60.core.model.LANParser
import com.openu60.core.network.AgentClient
import com.openu60.core.network.AgentError
import com.openu60.core.network.AuthManager
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class LANSettingsState(
    val config: LANConfig = LANConfig.empty,
    val isLoading: Boolean = false,
    val message: String? = null,
    val messageIsError: Boolean = false,
)

@HiltViewModel
class LANSettingsViewModel @Inject constructor(
    private val agentClient: AgentClient,
    private val authManager: AuthManager,
) : ViewModel() {

    private val _state = MutableStateFlow(LANSettingsState())
    val state: StateFlow<LANSettingsState> = _state.asStateFlow()

    fun updateConfig(config: LANConfig) {
        _state.value = _state.value.copy(config = config)
    }

    fun refresh() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val data = agentClient.getJSON("/api/network/lan")
                val config = LANParser.parse(data)
                _state.value = _state.value.copy(config = config, isLoading = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) refresh() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    fun save() {
        viewModelScope.launch {
            _state.value = _state.value.copy(isLoading = true, message = null)
            try {
                val c = _state.value.config
                val params = mapOf(
                    "lan_ipaddr" to c.lanIP,
                    "lan_netmask" to c.netmask,
                    "dhcp_enable" to if (c.dhcpEnabled) "1" else "0",
                    "dhcp_start" to c.dhcpStart,
                    "dhcp_end" to c.dhcpEnd,
                    "dhcp_lease_time" to c.dhcpLeaseTime,
                )
                agentClient.putJSON("/api/network/lan", params)
                _state.value = _state.value.copy(isLoading = false, message = "LAN settings saved", messageIsError = false)
            } catch (e: AgentError.Unauthorized) {
                if (authManager.reauthenticate()) save() else setError(e.message)
            } catch (e: Exception) {
                setError(e.message)
            }
        }
    }

    private fun setError(msg: String?) {
        _state.value = _state.value.copy(isLoading = false, message = msg ?: "Unknown error", messageIsError = true)
    }
}

File: mobile/ios/OpenU60/Features/RouterSettings/LAN/LANSettingsViewModel.swift
import SwiftUI

@Observable
@MainActor
final class LANSettingsViewModel {
    var config: LANConfig = .empty
    var isLoading: Bool = false
    var message: String?
    var messageIsError: Bool = false

    // Editable fields
    var editLanIP: String = ""
    var editNetmask: String = ""
    var editDhcpEnabled: Bool = false
    var editDhcpStart: String = ""
    var editDhcpEnd: String = ""
    var editLeaseTime: String = ""

    private let client: AgentClient
    private let authManager: AuthManager

    init(client: AgentClient, authManager: AuthManager) {
        self.client = client
        self.authManager = authManager
    }

    func refresh() async {
        isLoading = true
        message = nil

        do {
            let data = try await client.getJSON("/api/router/lan")
            config = LANParser.parse(data)
            syncEditFields()
        } catch {
            showMessage("Failed to load LAN: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    func apply() async {
        guard !editLanIP.isEmpty else {
            showMessage("LAN IP is required", isError: true)
            return
        }

        isLoading = true

        do {
            let _ = try await client.putJSON("/api/router/lan", body: [
                "lan_ipaddr": editLanIP,
                "lan_netmask": editNetmask,
                "dhcp_enable": editDhcpEnabled ? "1" : "0",
                "dhcp_start": editDhcpStart,
                "dhcp_end": editDhcpEnd,
                "dhcp_lease_time": editLeaseTime
            ])
            showMessage("LAN settings updated", isError: false)
            config = LANConfig(lanIP: editLanIP, netmask: editNetmask, dhcpEnabled: editDhcpEnabled,
                               dhcpStart: editDhcpStart, dhcpEnd: editDhcpEnd, dhcpLeaseTime: editLeaseTime)
        } catch {
            showMessage("Failed: \(error.localizedDescription)", isError: true)
        }

        isLoading = false
    }

    private func syncEditFields() {
        editLanIP = config.lanIP
        editNetmask = config.netmask
        editDhcpEnabled = config.dhcpEnabled
        editDhcpStart = config.dhcpStart
        editDhcpEnd = config.dhcpEnd
        editLeaseTime = config.dhcpLeaseTime
    }

    private func showMessage(_ text: String, isError: Bool) {
        message = text
        messageIsError = isError
    }
}
